nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/nar/NarRestApiClient.java

```diff
@@ -0,0 +1,138 @@
+import org.apache.nifi.web.client.api.HttpResponseStatus;
+import org.apache.nifi.web.client.api.WebClientService;
+import org.apache.nifi.web.client.api.WebClientServiceException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.nio.charset.StandardCharsets;
+import java.util.Objects;
+
+/**
+ * Encapsulate API calls for listing and downloading NARs through the REST API of a NiFi node.
+ */
+public class NarRestApiClient {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(NarRestApiClient.class);
+
+    private static final String HTTP_SCHEME = "http";
+    private static final String HTTPS_SCHEME = "https";
+
+    private static final String NIFI_API_PATH = "nifi-api";
+    private static final String CONTROLLER_PATH = "controller";
+    private static final String NAR_MANAGER_PATH = "nar-manager";
+    private static final String NARS_PATH = "nars";
+    private static final String NAR_DOWNLOAD_PATH = "download";
+
+    private final URI baseUri;
+    private final WebClientService webClientService;
+    private final ObjectMapper objectMapper;
+
+    public NarRestApiClient(final WebClientService webClientService, final String host, final int port, final boolean secure) {
+        try {
+            this.baseUri = new URI(secure ? HTTPS_SCHEME : HTTP_SCHEME, null, host, port, null, null, null);
+        } catch (URISyntaxException e) {
+            throw new RuntimeException(e);
+        }
+        this.webClientService = Objects.requireNonNull(webClientService, "WebClientService is required");
+        this.objectMapper = new ObjectMapper();
+        this.objectMapper.setAnnotationIntrospector(new JakartaXmlBindAnnotationIntrospector(objectMapper.getTypeFactory()));
+    }
+
+    public NarSummariesEntity listNarSummaries() {
+        final URI requestUri = new StandardHttpUriBuilder()
+                .scheme(baseUri.getScheme())
+                .host(baseUri.getHost())
+                .port(baseUri.getPort())
+                .addPathSegment(NIFI_API_PATH)
+                .addPathSegment(CONTROLLER_PATH)
+                .addPathSegment(NAR_MANAGER_PATH)
+                .addPathSegment(NARS_PATH)
+                .build();
+        LOGGER.info("Requesting NAR summaries from {}", requestUri);
+
+        final HttpRequestBodySpec requestBodySpec = webClientService.get()
+                .uri(requestUri)
+                .header("Accept", "application/json");
+
+        try (final HttpResponseEntity response = requestBodySpec.retrieve()) {
+            final InputStream responseBody = getResponseBody(requestUri, response);
+            return objectMapper.readValue(responseBody, NarSummariesEntity.class);
+        } catch (final WebClientServiceException | IOException e) {
+            throw new NarRestApiRetryableException(e.getMessage(), e);
```

Comment:
```suggestion
            throw new NarRestApiRetryableException("List NAR summaries failed", e);
```

Reply:
I wanted to leave this one as is because the message from the retryable exception will be logged in the retry loop in `StandardNarManager`, so it looks like this currently:
```
2024-07-03 16:46:27,810 INFO [main] org.apache.nifi.nar.StandardNarManager Synchronizing NARs with cluster coordinator
2024-07-03 16:46:27,914 WARN [main] org.apache.nifi.nar.StandardNarManager Error calling https://localhost:8443/nifi-api/controller/nar-manager/nars - 409 - The Flow Controller is initializing the Data Flow.: retrying
2024-07-03 16:46:27,914 INFO [main] org.apache.nifi.nar.StandardNarManager Unable to retrieve NAR summaries from cluster coordinator: retrying until [2024-07-03T20:51:27.811Z]
```
If we changed it as suggested, then the second line there would just say `List NAR summaries failed` with no info about why.

Reply:
Thanks for providing more context, that makes sense in light of the example described.